core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor.utils.js

```diff
@@ -0,0 +1,517 @@
+const oppiaResponseSelector = '.oppia-click-to-start-editing';
+const firstVersionOptionsButton = '#dropdownMenuButton-0';
+const secondVersionOptionsButton = '#dropdownMenuButton-1';
+
+
+// Preview tab elements.
+const testContinueButton = '.e2e-test-continue-button';
+const testFloatFormInput = '.e2e-test-float-form-input';
+const nextCardButton = '.e2e-test-next-card-button';
+const submitAnswerButton = '.e2e-test-submit-answer-button';
+const explorationRestartButton = '.e2e-preview-restart-button';
+const explorationConversationContent = '.e2e-test-conversation-content';
+
+// Settings tab elements.
+const explorationTitleInput = '.e2e-test-exploration-title-input';
+
+module.exports = class explorationAdmin extends baseUser {
+  /**
+  * Function for navigating to the contributor dashboard page.
+  */
+  async navigateToCreatorDashboard() {
+    await this.goto(creatorDashboardAdminUrl, { waitUntil: 'networkIdle0' });
+  }
+
+  /**
+  * Function for creating an exploration in the Exploration Editor
+  * @param {string} text - the text in the Exploration.
+  */
+  async createExploration(text, interaction) {
+    await this.clickOn(createExplorationButtonSelector);
+    await this.page.waitForSelector(
+      dismissWelcomeModalSelector, { visible: true });
+    await this.clickOn(dismissWelcomeModalSelector);
+    await this.page.waitForTimeout(300);
+    await this.clickOn(stateEditSelector);
+    await this.page.waitForSelector(
+      explorationTextInput, { visible: true });
+    await this.type(explorationTextInput, `${text}`);
+    await this.clickOn(saveContentButton);
+
+    await this.page.waitForSelector(addInteractionButton);
+    await this.clickOn(addInteractionButton);
+    await this.page.waitForSelector(
+      endInteractionSelector, { visible: true });
+    await this.clickOn(interaction);
+    await this.clickOn(saveInteractionButton);
+  }
+
+  /**
+  * Function for saving an exploration draft.
+  */
+  async saveExplorationDraft() {
+    await this.clickOn(saveChangesButton);
+    await this.page.waitForSelector(
+      saveDraftButton, { visible: true });
+    await this.page.waitForTimeout(500);
+    await this.clickOn(saveDraftButton);
+  }
+
+  /**
+  * Function to make metadata changes in the fucntio
+  * @param {string} text - the text of the Exploration.
+  */
+  async makeMetaDataChanges(title) {
+    await this.page.waitForTimeout(200);
+    await this.clickOn(
+      navigateToSettingsTabButton, { waitUntil: 'networkidle0' });
+    await this.page.waitForTimeout(200);
+    await this.type(explorationTitleInput, title);
+    await this.clickOn(navigateToMainTabButton, { waitUntil: 'networkidle0' });
+  }
+
+  /**
+  * Function to create multiple revisions of the same Exploration.
+  * @param {string} text - the text of the Exploration.
+  * @param {number} numOfVersions - number of versions to created.
+  */
+  async createMultipleRevisionsOfTheSameExploration(text, numOfVersion) {
```

Comment:
Here's how to think about functions in this file -- have each top-level function here reflect common actions of the end user, E.g. saving an exploration, adding a question, etc. Each action should be describable in a single short phrase and be something an end user would naturally do.

If an end user would not naturally do "make 10 revisions of this exploration" then don't have that function here. That corresponds to the end user *choosing* to make 10 separate edits -- those choices should be reflected in the "user" layer of the tests. But if the user's path for an action is constrained (e.g. they want to add a piece of feedback and that path involves clicking X, typing Y, clicking Z) -- then it's fine to have that entire action as a function in this file.

Hope this gives a helpful heuristic about how to structure the test between the different layers.

Reply:
I don't know. What behaviour are you testing -- what is/are the behaviour(s) for the end user that you want to preserve? State it like this: "As an exploration creator, if I am looking at the history tab for an exploration that has XXX revisions, I should be able to ABC."

Reply:
@seanlip,
Let me rephrase my concern.

As an exploration creator, if I am looking at the history tab for an exploration
that has more than 10 revisions, I should be able to navigate to the next page of revisions when the paginator is set to display 10 revisions per page.

Here’s the behavior I want to verify:

1) If there are more than 10 revisions, the ‘next page’ (>) button should be enabled, allowing the user to navigate to the next page of revisions.
2) If there are fewer than 10 revisions, the ‘next page’ (>) button should be disabled as there are no additional pages to navigate to.

My question is: Would checking the state of the ‘next page’ (>) button (whether it’s enabled or disabled) be a sufficient test for this paginator functionality?
 Or should I include additional checks? If so, may you please suggest?

Reply:
I would say that your check is reasonable, but I would also add clicking on the button in the "enabled" case and verifying that the right set of results show up (because that is what a user would do).

Reply:
sure, will take care.